src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+                                                     const BlockRealize& block_realize,
+                                                     const LoopSubspaceCollector& collector,
+                                                     arith::Analyzer* analyzer) {
+  const Block& block = block_realize->block;
+  DiagnosticContext diag_ctx(DiagnosticContext::Default(mod));
+
+  Array<Array<arith::IterMark>> division =
+      arith::SubspaceDivide(block_realize->iter_values, collector.loop_var_domain,
+                            collector.inner_loop_vars, block_realize->predicate,
+                            /*require_bijective=*/false, analyzer, diag_ctx);
+
+  if (division.empty()) {
+    // If we can't do perfect subspace division, check if it is a trivial case of subspace division.
+    // In this case, we can still blockize.
+    division = TrivialSubspaceDivision(block->iter_vars, block_realize->iter_values,
+                                       collector.outer_loop_vars, collector.inner_loop_vars,
+                                       block_realize->predicate);
+  }
+  if (division.empty()) {
+    throw SubspaceNotDivisibleError(mod, GetRef<For>(collector.inner_loops.back()), block);
+  }
+  return division;
+}
+
+/*!
+ * \brief The binding extractor to compute the bindings of the outer and the inner blocks after
+ * blockize.
+ */
+class BlockizedBindingExtractor {
+ public:
+  /*!
+   * \brief Extract bindings for blockize.
+   * \param iter_vars The iter vars of the original inner block.
+   * \param division The result of the subspace division.
+   */
+  void ExtractBindings(const Array<IterVar>& iter_vars,
+                       const Array<Array<arith::IterMark>>& division) {
+    ICHECK(iter_vars.size() + 1 == division.size());
+    for (size_t i = 0; i < iter_vars.size(); ++i) {
+      const IterVar& iter_var = iter_vars[i];
+      const arith::IterMapExprNode* outer_binding =
+          division[i][0]->source.as<arith::IterMapExprNode>();
+      const arith::IterMapExprNode* inner_binding =
+          division[i][1]->source.as<arith::IterMapExprNode>();
+      ICHECK(outer_binding);
+      ICHECK(inner_binding);
+
+      // After computing the subspace division, bindings[i] can be written as
+      // outer_binding * inner_binding->extent + inner_binding
+      // The outer block will have binding: iter_outer -> outer_binding
+      // The inner block will have binding: iter_inner -> iter_outer * inner_binding->extent +
+      // inner_binding
+
+      if (is_one(division[i][1]->extent)) {  // IsOuter
+        // extract this iter var to outer block directly
+        outer_bindings.push_back(
+            arith::NormalizeIterMapToExpr(GetRef<arith::IterMapExpr>(outer_binding)));
+        outer_iter_vars.push_back(iter_var);
+      } else {
+        const IterVar outer_var(Range::FromMinExtent(0, division[i][0]->extent),
+                                iter_var->var.copy_with_suffix("o"), iter_var->iter_type);
```

Comment:
Looks like we don't need the `const` here

```suggestion
        IterVar outer_var(Range::FromMinExtent(0, division[i][0]->extent),
                                iter_var->var.copy_with_suffix("o"), iter_var->iter_type);
```